experimental/mrpc/plugin/policy/delay/src/engine.rs

```diff
@@ -0,0 +1,202 @@
+    }
+
+    fn handle_request(&mut self, request: Vec<u8>, _cred: UCred) -> Result<()> {
+        let request: control_plane::Request = bincode::deserialize(&request[..])?;
+
+        match request {
+            control_plane::Request::NewConfig() => {
+                self.config = DelayConfig {};
+            }
+        }
+        Ok(())
+    }
+}
+
+impl_vertex_for_engine!(DelayEngine, node);
+
+impl Decompose for DelayEngine {
+    fn flush(&mut self) -> Result<usize> {
+        let mut work = 0;
+        while !self.tx_inputs()[0].is_empty() || !self.rx_inputs()[0].is_empty() {
+            if let Progress(n) = self.check_input_queue()? {
+                work += n;
+            }
+        }
+        Ok(work)
+    }
+
+    fn decompose(
+        self: Box<Self>,
+        _shared: &mut SharedStorage,
+        _global: &mut ResourceCollection,
+    ) -> (ResourceCollection, DataPathNode) {
+        let engine = *self;
+        let mut collections = ResourceCollection::with_capacity(4);
+        collections.insert("config".to_string(), Box::new(engine.config));
+        (collections, engine.node)
+    }
+}
+
+impl DelayEngine {
+    pub(crate) fn restore(
+        mut local: ResourceCollection,
+        node: DataPathNode,
+        _prev_version: Version,
+    ) -> Result<Self> {
+        let config = *local
+            .remove("config")
+            .unwrap()
+            .downcast::<DelayConfig>()
+            .map_err(|x| anyhow!("fail to downcast, type_name={:?}", x.type_name()))?;
+        let mut var_probability = 0.2;
+
+        let engine = DelayEngine {
+            node,
+            indicator: Default::default(),
+            config,
+            var_probability,
+        };
+        Ok(engine)
+    }
+}
+
+impl DelayEngine {
+    async fn mainloop(&mut self) -> EngineResult {
+        loop {
+            let mut work = 0;
+            loop {
+                match self.check_input_queue()? {
+                    Progress(0) => break,
+                    Progress(n) => work += n,
+                    Status::Disconnected => return Ok(()),
+                }
+            }
+            self.indicator.set_nwork(work);
+            future::yield_now().await;
+        }
+    }
+}
+
+#[inline]
+fn materialize_nocopy(msg: &RpcMessageTx) -> &hello::HelloRequest {
+    let req_ptr = msg.addr_backend as *mut hello::HelloRequest;
+    let req = unsafe { req_ptr.as_ref().unwrap() };
+    return req;
+}
+
+impl DelayEngine {
+    fn check_input_queue(&mut self) -> Result<Status, DatapathError> {
+        use phoenix_common::engine::datapath::TryRecvError;
+
+        match self.tx_inputs()[0].try_recv() {
+            Ok(msg) => {
+                match msg {
+                    EngineTxMessage::RpcMessage(msg) => {
+                        if rand::random::<f32>() < self.var_probability {
+                            let duration = time::Duration::from_millis(100);
+                            thread::sleep(duration);
```

Comment:
I would not suggest using thread::sleep inside an Engine, because this will block the entire runtime (there are other Engines running on it).

The runtime executor currently can drive Rust Futures, so ideally you could use some asynchronous analog to `std::thread:sleep` such as [smol::Timer](https://docs.rs/smol/latest/smol/struct.Timer.html) or [tokio::time::sleep](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html). However, since we haven't implement IO Reactors for the executor, it's not full-fledged as tokio or other async runtime. A workaround would be to return `Ok(Progress(0))` here and check again if it has passed the duration next time polled by the runtime executor.